src/core/credrank/markovProcessGraph.test.js

```diff
@@ -154,6 +157,112 @@ describe("core/credrank/markovProcessGraph", () => {
+              };
+              const edge = personalAttributionGadget.markovEdge(
+                details,
+                parameters.beta * proportion.proportionValue
+              );
+              checkMarkovEdge(mpg, edge);
+            }
+          }
+        }
+      });
+      it("only returns the latest attribution for a recipient in an epoch", () => {
+        expect.assertions(2);
+        const attr = attributions[0];
+        const attribution0 = [attr];
+        const mpg = markovProcessGraph(attribution0);
+        const interval = intervals[0];
+        for (const recipient of attr.recipients) {
+          const details = {
+            epochStart: interval.startTimeMs,
+            fromParticipantId: attr.fromParticipantId,
+            toParticipantId: recipient.toParticipantId,
+          };
+          const intervalProportions = recipient.proportions.filter(
+            (p) => p.timestampMs <= interval.startTimeMs
+          );
+          const proportion = intervalProportions.length
+            ? intervalProportions[intervalProportions.length - 1]
+                .proportionValue
+            : 0;
+          const edge = personalAttributionGadget.markovEdge(
+            details,
+            parameters.beta * proportion
+          );
+          checkMarkovEdge(mpg, edge);
+        }
+      });
+      it("propagates the latest proportion into subsequent epochs", () => {
+        expect.assertions(4);
+        const mpg = markovProcessGraph(attributions);
+        const interval = intervals[1];
+        for (const attr of attributions) {
+          for (const recipient of attr.recipients) {
+            const details = {
+              epochStart: interval.startTimeMs,
+              fromParticipantId: attr.fromParticipantId,
+              toParticipantId: recipient.toParticipantId,
+            };
+            const intervalProportions = recipient.proportions.filter(
+              (p) => p.timestampMs <= interval.startTimeMs
+            );
+            const proportion = intervalProportions.length
+              ? intervalProportions[intervalProportions.length - 1]
+                  .proportionValue
+              : 0;
+            const edge = personalAttributionGadget.markovEdge(
+              details,
+              parameters.beta * proportion
+            );
+            checkMarkovEdge(mpg, edge);
+          }
+        }
+      });
+      it("filters out 0-weighted edge addresses", () => {
+        const mpg = markovProcessGraph(attributions);
+        const interval = intervals[1];
+        const attr = attributions[1];
+        const recipient = attr.recipients[0];
+        const details = {
+          epochStart: interval.startTimeMs,
+          fromParticipantId: attr.fromParticipantId,
+          toParticipantId: recipient.toParticipantId,
+        };
+        const containsEdge = Array.from(mpg.edges()).includes(
+          personalAttributionGadget.markovEdge(details, 0)
+        );
```

Comment:
This is probably the maximum amount of logic that would fit within best practice.